### ISSUE TITLE
Steps of parallel process

### ISSUE THREAD

**Issue by lijingbu** [bcgsc/abyss#135]
Hey, 

Just found out that with **np option** abyss-pe will run the ABYSS in parallel, but the mapping step abyss-map (map each library reads to assebmly-3.fa) is only using all CPUs on one node, while the rest of the nodes are not running anything. I read in the document that ABYSS is the only step parallel.  It would be much faster if the abyss-map step could also use multiple nodes, in the case of multiple libraries.

Also, if running on a single machine with the **j option**, the abyss-pe uses only 1 thread to read the reads from each library, which is also a waste of resources. Currently, I am waiting on this step, will report if the following steps use multiple threads. 

Anyone can provide guides if anything I am missing here in terms of utilizing all computational resources? 

Thanks.
Lijing

**Comment by sjackman** [comment]
Hi, Lijing. Yes, `ABYSS-P` is the only distributed step of ABySS using Open MPI, all other steps are parallelized on a single machine using OpenMP. The default value of `j` should be the number of threads available on the first machine. For example, if you use four machines each with 12 threads, the default value of `np` is 48 and the default value of `j` is 12. If for some reason the scheduler integration did not work for you and the value detected for `j` is incorrect, you may specify a value for `j`.

**Comment by jos4uke** [comment]
Hi, 
just a remark about ABySS parallelization when run on a single machine. To be able to run all steps in multithreaded mode, one needs to set **np** and **j** options to the same value on `abyss-pe` command.  
I made the mistake once believing **j** value would be set to the value of **np**, but it is not done automatically. 
- If one set only **np** value (>1), **j** value sticks to the default value 1. And all steps will run on 1 cpu, even though abyss-bloom-dbg command is set with **j** value corresponding to **np** value, and paired-end assembly will not be parallelized.      
- If one set only **j** value (>1), **np** value sticks to the default value 1. And all steps will also run on 1 cpu as described above.   

14 days 1cpu assembly could be done in less than 20h with 50cpu for a 400Mb genome.
 
hope this could help

**Comment by benvvalk** [comment]
Hi @jos4uke,

Fair point and thank you.

For single node jobs I completely agree -- `j` should ideally make equal to `np` (unless otherwise specified by the user).  However, for multinode MPI jobs this would result in surprising and undesired behaviour, because the multithreaded programs only run on a single machine, whereas the MPI process (`ABYSS-P`) can make use of the cores on all of the machines allocated for the assembly job.

**Comment by sjackman** [comment]
Hi, Joseph. Are you running ABySS within a scheduler, like SGE or SLURM? Outside of a scheduler, the value of `j` defaults to the value of `np`. See
https://github.com/bcgsc/abyss/blob/master/bin/abyss-pe#L138
If the environment variable `PE_HOSTFILE` is defined by your scheduler, it uses the number of slots available on the first machine allocated to the job, which could potentially be 1.
See https://github.com/bcgsc/abyss/blob/master/bin/abyss-pe#L135
You can use `abyss-pe --dry-run` to report the commands that ABySS would run, without actually running them, which is useful for troubleshooting this sort of issue.

**Comment by benvvalk** [comment]
Thanks, @sjackman!

**Comment by jos4uke** [comment]
Hi @benvvalk 
you are right, having an openmpi environment have big advantages to run very intensive assembly jobs and particularly with ABySS taking full advantage of it.
I surely have to check for my assembly if something went wrong because of mpi jobs ran on my single node machine.
As far as I can see, from the log I got this kind of warnings about OMP variables:  
> libgomp: Invalid value for environment variable OMP_THREAD_LIMIT  
> libgomp: Invalid value for environment variable OMP_NUM_THREADS  

If I echo'ed these variables from the interactive shell on the same node running ABySS, I got this:
> $ echo $OMP_NUM_THREADS  
1    
> $ echo $OMP_THREAD_LIMIT   
1  

Not sure if this is harmful for the assembly.

Hi @sjackman 
you guessed it right, I ran ABySS within SLURM scheduler.
I didn't check for this PE_HOSTFILE  variable or its equivalent in SLURM, SLURM_JOB_NODELIST, before running `abyss-pe`. But it seems not set when I echo'ed it from the interactive shell on the same node running ABySS.
I should tell you also I didn't issue directly `abyss-pe` command but from a snakemake workflow. I passed to snakemake the number of cores (--cores) I requested from SLURM. And then from the rule that actually run `abyss-pe` I set the number of threads to the number of cores.
Plus, ABySS (v2.0.1) was installed using conda, and it comes with its openmpi dependency (version 2.1.2). As mentioned above about OMP variables, didn't know if this was set by default by conda or not.    
I didn't know if there is some interference due to conda+snakemake (conda+snakemake+SLURM?).
Anyway, when **j** and **np** are set to the same value, the number of cores, `abyss-pe` ran from snakemake within SLURM, it seems to work like a charm or as I expected it to work.  
Will definitely tell you how the assembly turns out

Thanks both of you
